### PR TITLE
Optimizations derived from armfazh/flor-sidh-x64 project.

### PR DESCRIPTION
**Pull request by armfazh**
Hi cloudflare, 

This is PR contains some optimizations derived from [flor-sidh-x64](http://github.com/armfazh/flor-sidh-x64) project. Basically, this PR shows the impact of the right-to-left algorithm in the calculation of shared secret function. Also, it contains a faster formula for computing point triplings, i.e. given a point P computes [3]P.

Please consider this PR as an improvement on the performance of your Go implementation.
Best,

---
### General optimizations

The following are general and well-known optimizations.

1. Replace ADD+DOUBLE by a xDBLADD function.
  This reduces 1M per iteration in 3-point ladder and 2M per iteration in ScalarMult function.
2. Print for types (useful for debugging purposes)

----
### Optimizations derived from FLOR-SIDH-x64
The following are specific optimizations based on the FLOR-SIDH-x64 work.

1. New formulas for point tripling.
2. New R2L method for shared secret, replaces 3-point ladder.

----
## Benchmark Comparison
For Shared Secret computation, the execution time was reduced by 9.5%.

| benchmark                         |   old ns/op  |   new ns/op   |  delta |
|-----------------------------------|--------------|---------------|--------|
| BenchmarkAliceKeyGen-4            |   33945573   |   32551477    | -4.11% |
| BenchmarkAliceKeyGenSlow-4        |   289292459  |   286781846   | -0.87% |
| BenchmarkBobKeyGen-4              |   37633036   |   37420870    | -0.56% |
| BenchmarkBobKeyGenSlow-4          |   472687918  |   465635415   | -1.49% |
| BenchmarkSharedSecretAlice-4      |   32414528   |   29297806    | **-9.62%** |
| BenchmarkSharedSecretAliceSlow-4  |   287817189  |   287222653   | -0.21% |
| BenchmarkSharedSecretBob-4        |   37216521   |   33594430    | **-9.73%** |
| BenchmarkSharedSecretBobSlow-4    |   473442424  |   464861983   | -1.81% |

----

### CODE REVIEW

File: optimizations.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+
```
Comment (vkrasnov) [inline optimizations.md]: This file should not be part of the PR

File: p751toolbox/curve.go [hunk 2]
```diff
@@ -226,6 +239,76 @@ func (xQ *ProjectivePrimeFieldPoint) Double(xP *ProjectivePrimeFieldPoint, aPlus
 	return xQ
 }
 
+// Given xP = x(P), xQ = x(Q), and xPmQ = x(P-Q), compute xPaddQ = x(P+Q) and  x2P = x(2P).
+func xDblAdd(curve *CachedCurveParameters, xP, xQ, xPmQ *ProjectivePoint) (x2P, xPaddQ ProjectivePoint) {
```
Comment (vkrasnov) [inline p751toolbox/curve.go]: I think that should probably be a method of *CachedCurveParameters

File: p751toolbox/curve.go [hunk 3]
```diff
@@ -226,6 +239,76 @@ func (xQ *ProjectivePrimeFieldPoint) Double(xP *ProjectivePrimeFieldPoint, aPlus
...
+	x2P.X.Mul(&AA, &BB)           // x4 = AA*(4C)*BB
+	x2P.Z.Mul(&curve.Aplus2C, &E) // z4 = (A+2C)*E
+	x2P.Z.Add(&x2P.Z, &BB)        // z4 = (4C)*BB+(A+2C)*E
+	x2P.Z.Mul(&x2P.Z, &E)         // z4 = E*((4C)*BB+(A+2C)*E)
+	return
+}
+
+// Given xP = x(P), xQ = x(Q), and xPmQ = x(P-Q), compute xPaddQ = x(P+Q) and  x2P = x(2P).
+// Assumes that the Z-xoordinate of PmQ is equal to 1.
+func xDblAdd_primefield(aPlus2Over4 *PrimeFieldElement, xP, xQ, xPmQ *ProjectivePrimeFieldPoint) (x2P, xPaddQ ProjectivePrimeFieldPoint) {
```
Comment (vkrasnov) [inline p751toolbox/curve.go]: Should be a method too

File: p751toolbox/curve.go [hunk 4]
```diff
@@ -226,6 +239,76 @@ func (xQ *ProjectivePrimeFieldPoint) Double(xP *ProjectivePrimeFieldPoint, aPlus
...
+	var A, AA, B, BB, C, D, E, DA, CB, t0, t1 PrimeFieldElement
```
Comment (vkrasnov) [inline p751toolbox/curve.go]: You could reuse temporary variables more efficiently, instead having so many of them

File: p751toolbox/curve.go [hunk 5]
```diff
@@ -226,6 +239,76 @@ func (xQ *ProjectivePrimeFieldPoint) Double(xP *ProjectivePrimeFieldPoint, aPlus
...
+
+	AA.Square(&A) // AA = A^2
+	BB.Square(&B) // BB = B^2
+
+	E.Sub(&AA, &BB) // E = AA-BB
+	DA.Mul(&D, &A)  // DA = D*A
+	CB.Mul(&C, &B)  // CB = C*B
+
+	t1.Add(&DA, &CB) // t1 = DA+CB
+	t0.Sub(&DA, &CB) // t0 = DA-CB
```
Comment (vkrasnov) [inline p751toolbox/curve.go]: For example reuse DA, instead t0, or even right into z5 (xPaddQ.X?) to avoid copying.

File: p751toolbox/curve.go [hunk 6]
```diff
@@ -226,6 +239,76 @@ func (xQ *ProjectivePrimeFieldPoint) Double(xP *ProjectivePrimeFieldPoint, aPlus
...
+	t1.Square(&t1)   // t1 = t1^2
+	t0.Square(&t0)   // t0 = t0^2
+
+	xPaddQ.X = t1         // z5 = z1*t1
```
Comment (vkrasnov) [inline p751toolbox/curve.go]: not sure if z and x in comments are mixed, or is it on purpose?

File: p751toolbox/curve.go [hunk 7]
```diff
@@ -452,6 +536,38 @@ func (xR *ProjectivePoint) ThreePointLadder(curve *ProjectiveCurveParameters, xP
 	return xR
 }
 
+/**
+Update: This is the right-to-left method for computing the x-coordinate of P+[k]Q.
```
Comment (vkrasnov) [inline p751toolbox/curve.go]: Update is not a valid comment IMO. I would modify the existing description.

File: p751toolbox/curve.go (hunk 4 above)
Comment (vkrasnov) [reply p751toolbox/curve.go]: Also applies to xDblAdd

File: p751toolbox/print.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
+package p751toolbox
+
+import (
+	"reflect"
```
Comment (vkrasnov) [inline p751toolbox/print.go]: You don't need reflect here. Please avoid it at all costs.

File: p751toolbox/print.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+			}
+			buffer.WriteString("\n")
+	}
+	return buffer.String()
+}
+
+func (element Fp751Element) String() string {
+	var out [94]byte
+	element.toBytesFromMontgomeryForm(out[:])
+	return fmt.Sprintf("%s",PrintHex(out[:]))
```
Comment (vkrasnov) [inline p751toolbox/print.go]: Inconsistent commas. Did you apply gofmt?

File: optimizations.md (hunk 1 above)
Comment (armfazh) [reply optimizations.md]: Agree.

File: p751toolbox/curve.go (hunk 2 above)
Comment (armfazh) [reply p751toolbox/curve.go]: Since this function operates over points, thus now it is a method of `ProjectivePoint` and `ProjectivePrimeFieldPoint`.
Good suggestion.

File: p751toolbox/curve.go (hunk 4 above)
Comment (armfazh) [reply p751toolbox/curve.go]: Now `DblAdd` uses more efficiently the auxiliary variables.

File: p751toolbox/curve.go (hunk 6 above)
Comment (armfazh) [reply p751toolbox/curve.go]: Fixed!

File: p751toolbox/curve.go (hunk 7 above)
Comment (armfazh) [reply p751toolbox/curve.go]: Agree with that.

File: p751toolbox/print.go (hunk 8 above)
Comment (armfazh) [reply p751toolbox/print.go]: reflect package dependency removed.

**Comment by armfazh** [comment]
In [4ec6a08](https://github.com/cloudflare/p751sidh/pull/2/commits/4ec6a08fc4a0137f410500b8e9a612ba4ffd2225) you can find the modifications derived from @vkrasnov  review.
